src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -318,3 +379,227 @@ ExecuteSelectTasksIntoTupleStore(List *taskList, TupleDesc resultDescriptor)
 
 	return resultStore;
 }
+
+
+/*
+ * ColocateFragmentsWithRelation moves the fragments in the cluster so they are
+ * colocated with the shards of target relation. These transfers are done by
+ * calls to fetch_intermediate_results() between nodes.
+ *
+ * returnValue[shardIndex] is list of result Ids that are colocated with
+ * targetRelation->sortedShardIntervalArray[shardIndex] after fetch tasks are
+ * done.
+ */
+static List **
+ColocateFragmentsWithRelation(List *fragmentList, DistTableCacheEntry *targetRelation)
+{
+	List *fragmentListTransfers = ColocationTransfers(fragmentList, targetRelation);
+	List *fragmentTransferTaskList = FragmentTransferTaskList(fragmentListTransfers);
+
+	ExecuteFetchTaskList(fragmentTransferTaskList);
```

Comment:
There is a question of whether we want/need to be able to tolerate failure on the write path. In that case, we may want to have fetch tasks with multiple placements, but that seems complex and requires us to have different queries for different placements.

My inclination would be to opt for simplicity and require all target placements to be available during the insert..select.

Reply:
> There is a question of whether we want/need to be able to tolerate failure on the write path. In that case, we may want to have fetch tasks with multiple placements, but that seems complex and requires us to have different queries for different placements.
> 
> My inclination would be to opt for simplicity and require all target placements to be available during the insert..select (keeping it the way it is now).

going to look into this today and tomorrow to see if there is a simple solution.